### PR TITLE
Convert subscriber.js into Typescript

### PR DESCRIPTION
**Pull request by codemist**
# References:

Jira: MNTOR-3100

Figma:

# Description

# Screenshot (if applicable)

Not applicable.

# How to test

# Checklist (Definition of Done)

- [ ] Localization strings (if needed) have been added.
- [ ] Commits in this PR are minimal and [have descriptive commit messages](https://chris.beams.io/posts/git-commit/).
- [ ] I've added or updated the relevant sections in readme and/or code comments
- [ ] I've added a unit test to test for potential regressions of this bug.
- [ ] If this PR implements a feature flag or experimentation, the Ship Behind Feature Flag status in Jira has been set
- [ ] Product Owner accepted the User Story (demo of functionality completed) or waived the privilege.
- [ ] All acceptance criteria are met.
- [ ] Jira ticket has been updated (if needed) to match changes made during the development process.
- [ ] Jira ticket has been updated (if needed) with suggestions for QA when this PR is deployed to stage.

### CODE REVIEW

File: src/db/tables/subscribers.ts [hunk 1]
```diff
@@ -390,41 +406,54 @@ async function getPotentialSubscribersWaitingForFirstDataBrokerRemovalFixedEmail
...
   if (!flag?.is_enabled) {
     return [];
   }
 
-  let query = knex('subscribers')
+  let query = knex("subscribers")
     .select()
     // Only send to users who haven't opted out of the monthly activity email...
-    .where((builder) => builder.whereNull("monthly_monitor_report").orWhere("monthly_monitor_report", true))
+    // eslint-disable-next-line @typescript-eslint/no-misused-promises
```
Comment (Vinnl) [inline src/db/tables/subscribers.ts:434]: `eslint-disable` comments are surprising, so we can help future readers by explaining it. Same for the one on line 441.

```suggestion
    // (It looks like Knex's `.where` type definition doesn't accept Promise-returning
    // functions, even though the code does; hence the `eslint-disable`)
    // eslint-disable-next-line @typescript-eslint/no-misused-promises
```

File: src/db/tables/subscribers.ts [hunk 2]
```diff
@@ -455,18 +484,18 @@ async function getSubscribersWaitingForMonthlyEmail (options = {}) {
  */
 // Not covered by tests; mostly side-effects. See test-coverage.md#mock-heavy
 /* c8 ignore start */
-async function updateMonthlyEmailTimestamp (email) {
-  const res = await knex('subscribers')
+async function updateMonthlyEmailTimestamp(email: string) {
```
Comment (Vinnl) [inline src/db/tables/subscribers.ts:487]: Nit, but might as well:

```suggestion
async function updateMonthlyEmailTimestamp(email: SubscriberRow["primary_email"]) {
```

File: src/db/tables/subscribers.ts [hunk 3]
```diff
@@ -579,32 +620,32 @@ async function getSubscribersWithUnresolvedBreaches (limit = 0) {
...
-
 /**
  * Private
  *
  * @param {any} subscriber
  */
 // Not covered by tests; mostly side-effects. See test-coverage.md#mock-heavy
 /* c8 ignore start */
-async function joinEmailAddressesToSubscriber (subscriber) {
+async function joinEmailAddressesToSubscriber(subscriber: SubscriberRow) {
```
Comment (Vinnl) [inline src/db/tables/subscribers.ts:638]: The `@param` above is now redundant.

File: src/knex-tables.d.ts [hunk 4]
```diff
@@ -131,7 +131,7 @@ declare module "knex/types/tables" {
     monthly_email_optout: boolean;
     monthly_monitor_report_at: null | Date;
     monthly_monitor_report: boolean;
-    breach_resolution: BreachResolution;
+    breach_resolution?: BreachResolution;
```
Comment (Vinnl) [inline src/knex-tables.d.ts:134]: Hmm, what was the reason behind this change? Inspecting my database, this appears to be a non-nullable column?

**Review by Vinnl — COMMENTED** [review]
Great steps forward for our codebase again! I'm a bit unsure about the `JSON.stringify` removals and there are still build errors, but otherwise mostly left some general feedback.

File: src/db/tables/subscribers.ts [hunk 5]
```diff
@@ -3,114 +3,109 @@
...
 /* c8 ignore start */
-async function getSubscriberByFxaUid (uid) {
-  const [subscriber] = await knex('subscribers').where({
-    fxa_uid: uid
-  })
-  const subscriberAndEmails = await joinEmailAddressesToSubscriber(subscriber)
-  return subscriberAndEmails
+async function getSubscriberByFxaUid(
+  uid: SubscriberRow["fxa_uid"],
+): Promise<SubscriberRow | null> {
```
[the remark this answers is not in the document]
Comment (Vinnl) [reply src/db/tables/subscribers.ts:38]: I would expect `SubscriberRow | undefined`, i.e. similar to what it was through JSDoc. Or did you see it being `any`?

**Review by rhelmer — APPROVED** [review]
lgtm outside of things @Vinnl already mentioned. Looks like the unit test failure is just a missing `jest.mock("../../db/tables/subscribers.js"` -> `jest.mock("../../db/tables/subscribers"`

**Pull request merged**